content/blog/2020-11-30-super-optimization/index.md

```diff
@@ -0,0 +1,49 @@
++++
+title = "Super Optimization"
+[extra]
+bio = """
+  Socrates Wong is a secound year PhD student at Cornell who intreset lies in Computer Archicture.
+"""
+latex = true
+[[extra.authors]]
+name = "Socrates Wong"
++++
+
+## Background
+Most complier optimization (peephole, dead code elimination, constant propagation, loop invariant code motion, common sub expression elimination) focus on transforming the program in a systematic manner, and the step taken by the optimizer are backed by formal rigorous proofs that are both sound and complete.  Superoptimizer is a new radical way of preforming optimization.  Instead of taking the traditional approach optimizers of transforming the input program, superoptimizer converts the problem in to a search space and attempts to find the shortest program that is able to provide the equivalent output.
+The paper we are discussing this week is ["Superoptimizer"][superopt] – A look at the Smallest Program” (Henry), this paper explains the purpose of superoptimizer, which converts an optimizing problem to a search space, how to reduce the search space and comparison with other common optimization techniques that are common at the publication of the paper. 
+
+
+## Super Optimization
+Super Optimization is finding the shortest program by exhaustive search inside a finite search space of the instruction set architecture (ISA).  Compare to other optimization techniques that focus on transformation and match, super optimization takes it to a whole different level.  Most optimizations techniques focus on the potential performance benefit on utilizing a transformation and ensuring correctness of the output by having proofs that the transformation does not affect the equivalence of the original program.  In modern compliers, this can ensure that the result is unchanged, but does not guarantee optimality of the output.  To address the need for the most optimal output, super optimization searches the entire sample spaces of the instruction set architecture for programs by generating all ombinations of eah instruction with various lengths, starting with the shortestlength.  
+
+Then super optimization then preform a test of the generated program determine if the generated program is equivalent with the orgnial program and if they are the same, then generated programwill be the optimal porgram for the ISA.  Although this soulds simple, verifing the generated program is a non trival task.  One of the sloution proposed in the paper was boolean test.  However as the complexity of some instructions have expressio of the order of 2^31 minterms (Massalin), it has caused a constriant on the number of instrcitios that can be generated and verified in a reasonable time to be limited to 3 instructions at the time of publication.  Therefore although the boolean test is intuitive, and correct a alternative test will be needed in order to test programs wiht larger amount of instructions.  
```

Comment:
What is a "minterm"? Probably needs explanation. Overall, I'm not sure what the sentence "However as the complexity of some instructions have expressio of the order of 2^31 minterms" means (centrally, I'm not sure what "complexity" means in the context means).